File: go-backend-interview-prep/05_errors_and_testing/05_panic_recover_vs_errors/main.go
package main

import (
	"errors"
	"fmt"
)

// safeDivide recovers from a panic inside a deferred function.
func safeDivide(a, b int) (result int, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("recovered panic: %v", r)
		}
	}()
	// integer division by zero panics in Go
	return a / b, nil
}

// mustPositive panics on programmer error (precondition violation).
func mustPositive(n int) int {
	if n <= 0 {
		panic(fmt.Sprintf("mustPositive: got %d, want > 0", n))
	}
	return n
}

// parseAge returns an error for expected bad input.
func parseAge(s string) (int, error) {
	var age int
	_, err := fmt.Sscanf(s, "%d", &age)
	if err != nil {
		return 0, fmt.Errorf("parseAge(%q): %w", s, err)
	}
	if age < 0 || age > 150 {
		return 0, errors.New("age out of range")
	}
	return age, nil
}

func main() {
	// --- Example 1: recover from panic ---
	result, err := safeDivide(10, 0)
	if err != nil {
		fmt.Println("safeDivide(10, 0):", err)
	} else {
		fmt.Println("safeDivide(10, 0):", result)
	}

	result, err = safeDivide(10, 3)
	fmt.Println("safeDivide(10, 3):", result)

	// --- Example 2: panic for programmer bugs ---
	fmt.Println("\nmustPositive(5):", mustPositive(5))

	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Println("caught panic:", r)
			}
		}()
		mustPositive(-1) // this panics
	}()

	// --- Example 3: errors for expected failures ---
	age, err := parseAge("25")
	fmt.Println("\nparseAge(\"25\"):", age, err)

	age, err = parseAge("abc")
	fmt.Println("parseAge(\"abc\"):", age, err)

	age, err = parseAge("999")
	fmt.Println("parseAge(\"999\"):", age, err)
}

File: go-backend-interview-prep/05_errors_and_testing/05_panic_recover_vs_errors/main.py
"""Panic, recover vs errors -- Python equivalent of the Go example."""


def safe_divide(a: int, b: int) -> int:
    """Catches ZeroDivisionError (like Go recover from panic)."""
    try:
        return a // b
    except ZeroDivisionError as e:
        raise ValueError(f"recovered: {e}") from e


def must_positive(n: int) -> int:
    """Panics (asserts) on programmer error."""
    assert n > 0, f"must_positive: got {n}, want > 0"
    return n


def parse_age(s: str) -> int:
    """Returns age or raises ValueError for bad input."""
    try:
        age = int(s)
    except ValueError:
        raise ValueError(f"parse_age({s!r}): not a number")
    if age < 0 or age > 150:
        raise ValueError("age out of range")
    return age


def main() -> None:
    # --- Example 1: catch crash-level error ---
    try:
        result = safe_divide(10, 0)
        print(f"safe_divide(10, 0): {result}")
    except ValueError as e:
        print(f"safe_divide(10, 0): {e}")

    print(f"safe_divide(10, 3): {safe_divide(10, 3)}")

    # --- Example 2: assertion for programmer bugs ---
    print(f"\nmust_positive(5): {must_positive(5)}")

    try:
        must_positive(-1)
    except AssertionError as e:
        print(f"caught assertion: {e}")

    # --- Example 3: exceptions for expected failures ---
    try:
        age = parse_age("25")
        print(f'\nparse_age("25"): {age} None')
    except ValueError as e:
        print(f'\nparse_age("25"): {e}')

    try:
        age = parse_age("abc")
        print(f'parse_age("abc"): {age} None')
    except ValueError as e:
        print(f'parse_age("abc"): 0 {e}')

    try:
        age = parse_age("999")
        print(f'parse_age("999"): {age} None')
    except ValueError as e:
        print(f'parse_age("999"): 0 {e}')


if __name__ == "__main__":
    main()
